File: 1396-count-servers-that-communicate/count-servers-that-communicate.c
int find_set(int x,int parent[]){
    if(parent[x]!=x){
        parent[x]=find_set(parent[x],parent);
    }
    return parent[x];
}
void union_set(int x, int y,int parent[],int rank[]){
    int p1=find_set(x,parent);
    int p2=find_set(y,parent);
    if(p1==p2){
        rank[p2]++;
        return;
    }
    if(rank[p1]<rank[p2]){
        parent[p1]=p2;
        rank[p2]+=rank[p1];
    }
    else{
        parent[p2]=p1;
        rank[p1]+=rank[p2];
    }
}
int countServers(int** grid, int gridSize, int* gridColSize){
    int m=gridSize;
    int n=*gridColSize;
    int parent[m*n];
    int rank[m*n];
    for(int i=0;i<m*n;i++){
        parent[i]=i;
        rank[i]=1;
    }
    //int ans=0;
    for(int i=0;i<m;i++){
        for(int j=0;j<n;j++){
            if(grid[i][j]==1){
                union_set(i,j+m,parent,rank);
            }
        }
    }
    int ans=0;
    for(int i=0;i<m*n;i++){
        if((parent[i]==i) && rank[i]>2){
            ans+=rank[i]-1;
        }
    }
    return ans;
}

File: 1396-count-servers-that-communicate/count-servers-that-communicate.py
class Solution:
    def countServers(self, grid: List[List[int]]) -> int:
        m=len(grid)
        n=len(grid[0])
        def find(x):
            if x!=parent[x]:
                parent[x]=find(parent[x])
            return parent[x]
        def union(x,y):
            p1,p2=find(x),find(y)
            if p1==p2:
                rank[p2]+=1
                return
            if rank[p1]<rank[p2]:
                parent[p1]=p2
                rank[p2]+=rank[p1]+1
            else:
                parent[p2]=p1
                rank[p1]+=rank[p2]+1
        parent=[i for i in range(m*n)] 
        rank=[0]*(m*n)
        for i in range(m):
            for j in range(n):
                if(grid[i][j]==1):
                    union(i,j+m)
        ans=0
        for i in range(m*n):
            if parent[i]==i and rank[i]>1:
                ans+=rank[i]
        return ans                    
